jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/internal/WebSocketChannel.java

```diff
@@ -614,59 +612,52 @@ public void onFrame(Frame frame, Callback callback)
+
+                // Handle inbound close
+                if (frame.getOpCode() == OpCode.CLOSE)
                 {
-                    // Handle inbound close
-                    if (frame.getOpCode() == OpCode.CLOSE)
+                    connection.cancelDemand();
+                    CloseStatus closeStatus = CloseStatus.getCloseStatus(frame);
+                    if (closed)
                     {
-                        connection.cancelDemand();
-                        CloseStatus closeStatus = CloseStatus.getCloseStatus(frame);
-                        if (state.onCloseIn(closeStatus))
-                        {
-                            callback = new Callback.Nested(callback)
-                            {
-                                @Override
-                                public void completed()
-                                {
-                                    handler.onClosed(state.getCloseStatus());
-                                    connection.close();
-                                }
-                            };
-                            handler.onFrame(frame, callback);
-                            return;
-                        }
-
                         callback = new Callback.Nested(callback)
                         {
                             @Override
                             public void completed()
                             {
-                                // was a close sent by the handler?
-                                if (state.isOutOpen())
-                                {
-                                    // No!
-                                    if (LOG.isDebugEnabled())
-                                        LOG.debug("ConnectionState: sending close response {}", closeStatus);
-
-                                    close(closeStatus.getCode(), closeStatus.getReason(), Callback.NOOP);
-                                    return;
-                                }
+                                handler.onClosed(channelState.getCloseStatus());
+                                connection.close();
                             }
                         };
+                        handler.onFrame(frame, callback);
+                        return;
                     }
 
-                    // Handle the frame
-                    handler.onFrame(frame, callback);
-                }
-                else
-                {
-                    if (LOG.isDebugEnabled())
-                        LOG.debug("Discarding post EOF frame - {}", frame);
-                    callback.failed(new EofException());
+                    callback = new Callback.Nested(callback)
+                    {
+                        @Override
+                        public void completed()
+                        {
+                            // was a close sent by the handler? todo what if someone beats this to close
```

Comment:
It is unlikely that an application will call the callback for receiving a close frame without closing, but then in another thread issue a close that races with the close below.  Perhaps we should be trapping ISE thrown by the close below and ignoring them... actually we do that already with the NOOP!  So I think we can replace this `todo` with an explanation.

Reply:
Many applications have their own threads for dealing with writes/events on websocket.
It's quite common to have race conditions here.

Reply:
@joakime really?   If we call the application with an `onFrame(closeFrame,callback)` AND during that call it does not send a close frame in response  AND prior to calling the `callback.succeeded()` it has still not sent a close frame in response,  yet it is still going to somehow organise another thread to send a close frame to race with the container?    I think that is highly unlikely.   Anyway, if they win the race, we ignore the error.  If we win the race, they will get a failure telling them the websocket was already closed, which is something that any application really needs to handle anyway.   So I think we are good here and just need to explain rather than TODO some future change.